### PR TITLE
feat(source): let shared SourceExecutor start from latest instead of specified offset

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
https://github.com/risingwavelabs/risingwave/issues/16576#issuecomment-2095571890

mentioned by Eric, this is also his original idea in his RFC

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/connector/src/source/kafka/split.rs [hunk 1]
```diff
@@ -24,6 +24,12 @@ pub struct KafkaSplit {
     pub(crate) partition: i32,
     pub(crate) start_offset: Option<i64>,
     pub(crate) stop_offset: Option<i64>,
+    #[serde(skip)]
+    /// Used by shared source to hackily seek to the latest offset without fetching start offset first.
+    /// XXX: But why do we fetch low watermark for latest start offset..?
```
Comment (BugenZhao) [inline src/connector/src/source/kafka/split.rs:29]: Does it mean that theoretically we can also follow this approach for the Kafka sources specified `scan.startup.mode='latest'` by users?

**Review by BugenZhao — COMMENTED** [review]
LGTM

Have you verified when the `Offset::End` is resolved? In other words, if you create a `Reader` and then immediately pause it (without polling once), will the reader eventually start from the latest offset resolved during creation or when the stream is resumed and first polled?

File: src/connector/src/source/kafka/split.rs [hunk 2]
```diff
@@ -24,6 +24,12 @@ pub struct KafkaSplit {
...
+    ///
+    /// When this is `true`, `start_offset` will be ignored.
+    pub(crate) hack_seek_to_latest: bool,
```
Comment (tabVersion) [inline src/connector/src/source/kafka/split.rs:32]: suggest changing another name like `force_start_latest`. This name sounds like doing some magic 🪄

**Review by tabVersion — APPROVED** [review]
LGTM, how about writing a new post about the shared source arch.

File: src/connector/src/source/kafka/split.rs (hunk 1 above)
Comment (xxchan) [reply src/connector/src/source/kafka/split.rs:29]: I do think so. @tabVersion so why do we fetch low watermark instead of passing Offset::End?

**Comment by xxchan** [comment]
> Have you verified when the `Offset::End` is resolved? 

It's when `consumer.assign` is called. So pausing doesn't change the starting position.

File: src/connector/src/source/kafka/split.rs (hunk 2 above)
Comment (xxchan) [reply src/connector/src/source/kafka/split.rs:32]: I do think this is hacky and magical.. If possibile, I want to only use `start_offset` to do this. In other words, also use `Offset::End` for `scan.startup.mode='latest'`

File: src/connector/src/source/kafka/split.rs (hunk 1 above)
Comment (xxchan) [reply src/connector/src/source/kafka/split.rs:29]: Note that pulsar uses `PulsarEnumeratorOffset` in `PulsarSplit`, so I think it's perfectly ok.

**Pull request merged**